### DISCUSSION TITLE
NsisBootstrapper with multiple prerequisites and MSI with Custom dialogs

### DISCUSSION THREAD

**Discussion by PaulH-HBK** [oleg-shilo/wixsharp#1710] [Q&A]
Having created an MSI with standard dialogs to install software we created a bootstrapper so we could include some dependencies such as .NET which went reasonably well.  We wanted to add a custom dialog into the installer so added this to the MSI and set the DisplayInternalUI flag.  The MSI shows the custom dialog fine but the bootstrapper does not show it.  After some reading in this group it appears a standard bootstrapper does not show custom dialogs and the NsisBootstrapper is required.  After looking into this it doesn't appear to be easy to have multiple prerequisites (we have 3).
At a bit of a loss now to decide the best way to proceed.
Looking for the best way to package an MSI with custom dialog and 3 prerequisites (all third party executables).
Any advice would be greatly appreciated.

**Comment by geghamyan** [comment]
Hi @PaulH-HBK ,
Please take a look on this reply and [discussion](https://github.com/oleg-shilo/wixsharp/discussions/1672#discussioncomment-11229410
).

I hope that helps,
Tigran

**Comment by geghamyan** [comment]
In addition - I believe Oleg has added the `MsiExePackage` class specifically to address situations where you need to display a custom UI of the MSI installer. Please take a look at `MsiExePackage` here: [GitHub Link](https://github.com/oleg-shilo/wixsharp/releases/tag/v2.2.0.0).

In general, the decision depends on the nature of the dependencies you are installing. If you need to handle them as a whole — where installing, uninstalling, or reinstalling likely applies to all packages in the bundle — the Burn bootstrapper might be a better choice. However, if the dependencies are more like prerequisites that you install on the machine and leave in place (even if your main MSI is later removed), the NSISBootstrapper is a much simpler option.

If you step away from the standard Burn bootstrapper workflows, such as creating a custom UI for the bootstrapper or a silent bootstrapper, it can become very challenging to ensure it behaves correctly in edge cases. I've spent countless hours troubleshooting such issues.

**Comment by BenniKo** [comment]
Hello @PaulH-HBK,
I have a similar problem.
Did you find a way to show a custom dialog (with bootstrap)?

**Reply by PaulH-HBK** [reply]
Apologies I have taken so long to respond.  Unfortunately we did not resolve this issue.

**Comment by oleg-shilo** [comment]
If you want to reliable show MSI UI without depending on WiX Burn then you can have a look at this sample:

<img width="294" height="193" alt="image" src="https://github.com/user-attachments/assets/c0037956-955f-4b82-b6cf-12a3f89722ca" />

Burn always suppresses MSI own UI. 
It's a long story. WiX3 had this bug for years. WiX4 fixed it, but later it was reintroduced again. So in that sample above, I showed how you can simply wrapp a given msi in the self-executable and then add it to the bundle as an exe package.

If you really-really want a custom bootstrapper, then look at the WixBootstrapper_UI sample. Read carefully the explanations. WiX does not make UI bootstrapper easy, but it's doable.
